// @openreplay/tracker/cjs/app/observer/top_observer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const observer_js_1 = require("./observer.js");
const iframe_observer_js_1 = require("./iframe_observer.js");
const shadow_root_observer_js_1 = require("./shadow_root_observer.js");
const index_js_1 = require("../../messages/index.js");
const attachShadowNativeFn = Element.prototype.attachShadow;
class TopObserver extends observer_js_1.default {
    constructor(app, options) {
        super(app, Object.assign({
            captureIFrames: false
        }, options));
        this.iframeObservers = [];
        this.shadowRootObservers = [];
        // IFrames
        this.app.nodes.attachNodeCallback(node => {
            if ((0, observer_js_1.isInstance)(node, HTMLIFrameElement) &&
                (this.options.captureIFrames || node.getAttribute("data-openreplay-capture"))) {
                this.handleIframe(node);
            }
        });
        // ShadowDOM
        this.app.nodes.attachNodeCallback(node => {
            if ((0, observer_js_1.isInstance)(node, Element) && node.shadowRoot !== null) {
                this.handleShadowRoot(node.shadowRoot);
            }
        });
    }
    handleIframe(iframe) {
        let context = null;
        const handle = this.app.safe(() => {
            const id = this.app.nodes.getID(iframe);
            if (id === undefined) {
                return;
            } //log
            if (iframe.contentWindow === context) {
                return;
            } //Does this happen frequently?
            context = iframe.contentWindow;
            if (!context) {
                return;
            }
            const observer = new iframe_observer_js_1.default(this.app, this.options, context);
            this.iframeObservers.push(observer);
            observer.observe(iframe);
        });
        this.app.attachEventListener(iframe, "load", handle);
        handle();
    }
    handleShadowRoot(shRoot) {
        const observer = new shadow_root_observer_js_1.default(this.app, this.options, this.context);
        this.shadowRootObservers.push(observer);
        observer.observe(shRoot.host);
    }
    observe() {
        // Protection from several subsequent calls?
        const observer = this;
        Element.prototype.attachShadow = function () {
            const shadow = attachShadowNativeFn.apply(this, arguments);
            observer.handleShadowRoot(shadow);
            return shadow;
        };
        // Can observe documentElement (<html>) here, because it is not supposed to be changing.
        // However, it is possible in some exotic cases and may cause an ignorance of the newly created <html>
        // In this case context.document have to be observed, but this will cause 
        // the change in the re-player behaviour caused by CreateDocument message: 
        //   the 0-node ("fRoot") will become #document rather than documentElement as it is now.
        // Alternatively - observe(#document) then bindNode(documentElement)
        this.observeRoot(this.context.document, () => {
            this.app.send(new index_js_1.CreateDocument());
        }, this.context.document.documentElement);
    }
    disconnect() {
        Element.prototype.attachShadow = attachShadowNativeFn;
        this.iframeObservers.forEach(o => o.disconnect());
        this.iframeObservers = [];
        this.shadowRootObservers.forEach(o => o.disconnect());
        this.shadowRootObservers = [];
        super.disconnect();
    }
}
exports.default = TopObserver;
